src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp

```diff
@@ -48,12 +55,147 @@ static MemoryMapRegionType WindowsProtectionToInternal(uint64_t protection) {
+        foreach2 (it, other.begin(), other.end()) { insert(it.start(), it.stop(), *it); }
+    }
+};
+
+MemoryMapRegion::Allocator MemoryMapRegion::s_alloc;
+
+///
+/// \brief Maintains a region map for each process id
+///
+typedef std::unordered_map<uint64_t, MemoryMapRegion> MemoryMapRegionProcess;
+
+class MemoryMapRegionManager {
+private:
+    MemoryMapRegionProcess m_regions;
+
+public:
+    void addRegion(uint64_t pid, uint64_t addr, uint64_t end, MemoryMapRegionType type) {
+        assert(end > addr);
+        removeRegion(pid, addr, end); // FIXME: make it faster
+
+        auto &map = m_regions[pid];
+        map.insert(addr, end - 1, type);
+    }
+
+    void removeRegion(uint64_t target_pid, uint64_t addr, uint64_t end) {
+        assert(end > addr);
+
+        MemoryMapRegion &map = m_regions[target_pid];
+
+        MemoryMapRegion::iterator ie = map.end();
+        MemoryMapRegion::iterator it;
+
+        while (((it = map.find(addr)) != ie) && (it.start() < end)) {
+            uint64_t it_addr = it.start();
+            uint64_t it_end = it.stop();
+            MemoryMapRegionType type = (*it);
+            it.erase();
+
+            if (it_addr < addr) {
+                map.insert(it_addr, addr - 1, type);
+            }
+            if (it_end > end - 1) {
+                map.insert(end, it_end, type);
+            }
+        }
+    }
+
+    MemoryMapRegionType lookupRegion(uint64_t pid, uint64_t addr) const {
+        MemoryMapRegionProcess::const_iterator it = m_regions.find(pid);
+        if (it == m_regions.end()) {
+            return MM_NONE;
+        }
+
+        return (*it).second.lookup(addr, MM_NONE);
+    }
+
+    bool lookupRegion(uint64_t pid, uint64_t addr, uint64_t &start, uint64_t &end, MemoryMapRegionType &type) const {
+        MemoryMapRegionProcess::const_iterator it = m_regions.find(pid);
+        if (it == m_regions.end()) {
+            return false;
+        }
+
+        const MemoryMapRegion *map = &(*it).second;
+        auto rit = map->find(start);
+        if (rit == map->end()) {
+            return false;
+        }
+
+        start = rit.start();
+        end = rit.stop();
+        type = *rit;
+        return true;
+    }
+
+    void iterateRegions(uint64_t pid, MemoryMapCb &callback) const {
+        MemoryMapRegionProcess::const_iterator it = m_regions.find(pid);
+        if (it == m_regions.end()) {
+            return;
+        }
+
+        // XXX: somehow c++11 for (auto it:map) doesn't work
+        const MemoryMapRegion &map = (*it).second;
+        foreach2 (rit, map.begin(), map.end()) {
+            if (!callback(rit.start(), rit.stop(), *rit)) {
+                break;
+            }
+        }
+    }
+
+    void removePid(uint64_t pid) {
+        MemoryMapRegionProcess::iterator it = m_regions.find(pid);
+        if (it != m_regions.end()) {
+            m_regions.erase(it);
+        }
+    }
+
+    void dump(llvm::raw_ostream &os) const {
```

Comment:
Again can we remove these `foreach2` macros?

Reply:
I replaced them with the normal for(;;) style. It's almost as short as foreach2.